File: app/src/main/java/com/friendlypos/reimpresion/activity/ReimprimirActivity.java
package com.friendlypos.reimpresion.activity;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;

import com.friendlypos.R;
import com.friendlypos.application.util.Functions;
import com.friendlypos.distribucion.activity.DistribucionActivity;
import com.friendlypos.distribucion.fragment.BaseFragment;
import com.friendlypos.distribucion.fragment.DistResumenFragment;
import com.friendlypos.distribucion.fragment.DistSelecClienteFragment;
import com.friendlypos.distribucion.fragment.DistSelecProductoFragment;
import com.friendlypos.distribucion.fragment.DistTotalizarFragment;
import com.friendlypos.distribucion.util.Adapter;
import com.friendlypos.principal.activity.BluetoothActivity;
import com.friendlypos.principal.activity.MenuPrincipal;
import com.friendlypos.reimpresion.fragment.ReimprimirFacturaFragment;
import com.friendlypos.reimpresion.fragment.ReimprimirResumenFragment;

import java.util.ArrayList;
import java.util.List;

public class ReimprimirActivity extends BluetoothActivity {
    private Toolbar toolbar;
    private TabLayout tabLayout;
    private ViewPager viewPager;

    private String invoiceIdReimprimir;
    private int selecFacturaTab;

    public String getInvoiceIdReimprimir() {
        return invoiceIdReimprimir;
    }

    public void setInvoiceIdReimprimir(String invoiceIdReimprimir) {
        this.invoiceIdReimprimir = invoiceIdReimprimir;
    }

    public int getSelecFacturaTab() {
        return selecFacturaTab;
    }

    public void setSelecFacturaTab(int selecFacturaTab) {
        this.selecFacturaTab = selecFacturaTab;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_reimprimir);
        toolbar = (Toolbar) findViewById(R.id.toolbarReimprimir);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        viewPager = (ViewPager) findViewById(R.id.viewpagerReimprimir);
        setupViewPager(viewPager);

        tabLayout = (TabLayout) findViewById(R.id.tabsReimprimir);
        tabLayout.setupWithViewPager(viewPager);
        tabLayout.setTabMode(TabLayout.MODE_FIXED);

        viewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabLayout));
        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {

            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                int tabCliente = getSelecFacturaTab();
                if (tabCliente == 0 && tab.getPosition() != 0) {

                    Functions.CreateMessage(ReimprimirActivity.this, "Distribución", "Seleccione una factura.");

                    new Handler().postDelayed(
                            new Runnable() {

                                @Override
                                public void run() {
                                    tabLayout.getTabAt(0).select();
                                }
                            }, 100);
                }
                else {

                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });

       /* tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {

            @Override
           public void onTabSelected(TabLayout.Tab tab) {
                if (tab.getPosition() != 0) {

                    Functions.CreateMessage(DistribucionActivity.this, "Distribución", "Seleccione una factura.");

                    new Handler().postDelayed(
                            new Runnable() {

                                @Override
                                public void run() {
                                    tabLayout.getTabAt(0).select();
                                }
                            }, 100);
                }
                else {

                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });*/
    }

  /*  private void setupViewPager(ViewPager viewPager) {
        Adapter adapter = new Adapter(getSupportFragmentManager());
        adapter.addFragment(new ReimprimirFacturaFragment(), "SELECCIONE LA FACTURA");
        adapter.addFragment(new ReimprimirResumenFragment(), "RESUMEN");
        //adapter.addFragment(new quickSaleTotalFragment(), "TOTALIZAR");
        viewPager.setAdapter(adapter);
    }*/

   private void setupViewPager(ViewPager viewPager) {
        Adapter adapter = new Adapter(getSupportFragmentManager());
        final List<BaseFragment> list = new ArrayList<>();
        list.add(new ReimprimirFacturaFragment());
        list.add(new ReimprimirResumenFragment());

        adapter.addFragment(list.get(0), "Seleccionar Factura");
        adapter.addFragment(list.get(1), "Resumen");

        viewPager.setAdapter(adapter);
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                list.get(position).updateData();
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:

                Intent intent = new Intent(ReimprimirActivity.this, MenuPrincipal.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(intent);
                finish();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }}
}

File: app/src/main/java/com/friendlypos/reimpresion/activity/ReimprimirActivity.kt
package com.friendlysystemgroup.friendlypos.reimpresion.activity

import android.app.ActivityManager
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.KeyEvent
import android.view.MenuItem
import android.view.WindowManager
import androidx.viewpager.widget.ViewPager
import com.friendlysystemgroup.friendlypos.application.bluetooth.PrinterService
import com.friendlysystemgroup.friendlypos.application.util.Functions.CreateMessage
import com.friendlysystemgroup.friendlypos.databinding.ActivityReimprimirBinding
import com.friendlysystemgroup.friendlypos.distribucion.fragment.BaseFragment
import com.friendlysystemgroup.friendlypos.distribucion.util.Adapter
import com.friendlysystemgroup.friendlypos.principal.activity.BluetoothActivity
import com.friendlysystemgroup.friendlypos.principal.activity.MenuPrincipal
import com.friendlysystemgroup.friendlypos.reimpresion.fragment.ReimprimirFacturaFragment
import com.friendlysystemgroup.friendlypos.reimpresion.fragment.ReimprimirResumenFragment
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.TabLayoutOnPageChangeListener

/**
 * Actividad para la reimpresión de facturas
 */
class ReimprimirActivity : BluetoothActivity() {
    private lateinit var binding: ActivityReimprimirBinding
    
    // Identificadores de la factura y estado de selección
    var invoiceIdReimprimir: String? = null
        private set
    private var selecFacturaTab: Int = 0
    
    // Lista de fragmentos
    private lateinit var fragmentList: List<BaseFragment>
    
    companion object {
        private const val TAG = "ReimprimirActivity"
        private const val TAB_DELAY_MS = 100L
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        
        binding = ActivityReimprimirBinding.inflate(layoutInflater)
        setContentView(binding.root)
        
        setupToolbar()
        setupViewPager()
        setupTabs()
        connectToPrinter()
        
        Log.d(TAG, "Actividad creada correctamente")
    }
    
    private fun setupToolbar() {
        setSupportActionBar(binding.toolbarReimprimir)
        supportActionBar?.apply {
            setDisplayShowHomeEnabled(true)
            setDisplayHomeAsUpEnabled(true)
        }
    }

    private fun setupViewPager() {
        val adapter = Adapter(supportFragmentManager)
        fragmentList = listOf(
            ReimprimirFacturaFragment.newInstance(),
            ReimprimirResumenFragment.newInstance()
        )
        
        adapter.addFragment(fragmentList[0], "Seleccionar Factura")
        adapter.addFragment(fragmentList[1], "Resumen")

        binding.viewpagerReimprimir.apply {
            this.adapter = adapter
            addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageScrolled(
                    position: Int,
                    positionOffset: Float,
                    positionOffsetPixels: Int
                ) { /* No se requiere acción */ }

                override fun onPageSelected(position: Int) {
                    fragmentList[position].updateData()
                }

                override fun onPageScrollStateChanged(state: Int) {
                    /* No se requiere acción */
                }
            })
        }
    }
    
    private fun setupTabs() {
        binding.apply {
            tabsReimprimir.setupWithViewPager(viewpagerReimprimir)
            tabsReimprimir.tabMode = TabLayout.MODE_FIXED
            
            viewpagerReimprimir.addOnPageChangeListener(
                TabLayoutOnPageChangeListener(tabsReimprimir)
            )
            
            tabsReimprimir.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
                override fun onTabSelected(tab: TabLayout.Tab) {
                    if (selecFacturaTab == 0 && tab.position != 0) {
                        mostrarMensajeSeleccioneFactura()
                    }
                }

                override fun onTabUnselected(tab: TabLayout.Tab) {
                    /* No se requiere acción */
                }

                override fun onTabReselected(tab: TabLayout.Tab) {
                    /* No se requiere acción */
                }
            })
        }
    }
    
    private fun mostrarMensajeSeleccioneFactura() {
        CreateMessage(
            this,
            "Reimpresión",
            "Seleccione una factura."
        )

        Handler(Looper.getMainLooper()).postDelayed({
            binding.tabsReimprimir.getTabAt(0)?.select()
        }, TAB_DELAY_MS)
    }

    private fun connectToPrinter() {
        if (printer_enabled) {
            printer?.takeIf { it.isNotEmpty() }?.let { printerAddress ->
                if (!isServiceRunning(PrinterService.CLASS_NAME)) {
                    PrinterService.startRDService(applicationContext, printerAddress)
                }
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                navigateToMainMenu()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun navigateToMainMenu() {
        Intent(this, MenuPrincipal::class.java).apply {
            addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
            startActivity(this)
        }
        finish()
    }

    private fun isServiceRunning(serviceClassName: String): Boolean {
        val activityManager = getSystemService(ACTIVITY_SERVICE) as ActivityManager
        val services = activityManager.getRunningServices(Int.MAX_VALUE)
        return services.any { it.service.className == serviceClassName }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            navigateToMainMenu()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }
    
    fun setSelectedInvoice(invoiceId: String) {
        invoiceIdReimprimir = invoiceId
        selecFacturaTab = 1
    }
}
